### ISSUE TITLE
Type 2 quantile incorrect for high order quantiles

### ISSUE THREAD

**Issue by mafavaron** [serv-terr/pbl_met#21]
The problem was made evident by testing procedure.

As apparent from test data, the _pbl_met_ and R version differ significantly for Type 2, when quantile is 0.9 (90%) or larger.

All other types (1 to 9, except 2 of course) work correctly, with differences from the R version in the order of the machine epsilon.

**Comment by mafavaron** [comment]
The sorted values of the test vector are listed in the following table:

Order stat    Value
       01          7.76116271E-03
       02         7.00317249E-02
       03         8.08336139E-02
       04         8.72771814E-02
       05         8.92919824E-02
       06         0.108502716
       07         0.227665439
       08        0.261844039
       09        0.328924268
       10         0.392417699
       11          0.397314131
       12          0.423296690
       13         0.449646443
       14         0.495459318
       15         0.544526339
       16         0.565658867
       17         0.576010168
       18         0.613014996
       19         0.622887552
       20        0.625548840
       21        0.639082491
       22        0.696100116
       23        0.710998237
       24        0.727310419
       25        0.741843641
       26        0.781121254
       27        0.819947362
       28        0.878841460
       29        0.902220607
       30        0.915510356
       31        0.974638939
       32        0.975520134

This may prove useful when appreciating results.

Now: for the 8-th element of type 2, the R estimate is 0.902220600750297, while the _pbl_met_ estimate is 0.908865452: quite a large difference.

Namely: R picked up the actual 29-th order statistic, and did not average. _pbl_met_ averaged order stats 29 and 30.

**Comment by mafavaron** [comment]
The following table compares values found at large quantiles.

Quantile level        Value (R)                           Value (_pbl_met_)           Note
    0.7500               0.734577046125196       0.734577060                  Avg, almost within mach.epsilon
    0.9000               0.902220600750297      0.908865452                  Pick vs Avg
    0.9500               0.974638933548704      0.975079536                  Pick vs Avg
    0.9980               0.975520152132958       0.975520134                   Pick, almost within mach.epsilon
    0.9999               0.975520152132958       0.975520134                   Pick, almost within mach.epsilon
    1.0000                0.975520152132958       0.975520134                   Pick, almost within mach.epsilon

In conclusion:

1) Type 2 implementation of R start picking at 0.9000, which in case of _pbl_met_ becomes 0.9980.

2) When both methods average, or pick, result is the same (up to machine epsilon).

3) When R picks, it does to the value closest to 'h', while _pbl_met_ always pick at order statistics 32.

Further investigation about the relationship between quantile level and value is quite worth.

**Comment by mafavaron** [comment]
In _pbl_met_, the integer index 'h' is computed as

    h = n * p + 0.5

where n is the sample size, and p the desired quantile level. This formula, incidentally, has been taken from Wikipedia, voice "quantile", making reference to the original paper [Hyndman, 1996].

The story is apparently much different with the R documentation, also referring to [Hyndman, 1996], but very closely following the original reference (which was then added to the manual, and to the collection of _pbl_met_ foundational papers).

Before to do wide-range investigations, it could be better to check the original paper and see which the actual formula for type 2 sample quantiles really is (and, in case of wrong Wikipedia formulation, would I have some spare time, fix it).

**Comment by mafavaron** [comment]
And in fact, using the original [Hyndman, 1996] formulae we get a completely different rule than the one written to date in Wikipedia.

By using the new `q.type2.test` function in the R test support script for pbl_stat, we get the following interesting table (comments added by me).
| Case | p = Q.level | j | g              | $$\gamma$$      |  Note                                                       |
|:--|-------|---|---------|----|----------------------------------|
| 1  |0.0000  | 0 |0.0000  | 0.5 | j < 1, then value at index 1 used |
| 2 | 0.0001  | 0 | 0.0032  | 1.0 | j < 1, then value at index 1 used |
| 3 | 0.0010  | 0 | 0.0320  | 1.0 | j < 1, then value at index 1 used |
| 4 | 0.0100  | 0 | 0.3200  | 1.0 | j < 1, then value at index 1 used |
| 5 | 0.1000  | 3 | 0.2000  | 1.0 | Picker, not averager |
| 6 | 0.5000 | 16 | 0.0000  | 0.5 |  |
| 7 | 0.7500 | 24 | 0.0000  | 0.5 |  |
| 8 | 0.9000 | 28 | 0.8000  | 1.0 | Picker, not averager |
| 9 | 0.9500 | 30 | 0.4000  | 1.0 | Picker, not averager |
| 10| 0.9980 | 31 | 0.9360  | 1.0 | Picker, not averager |
| 11 | 0.9999 | 31 | 0.9968  | 1.0 | Picker, not averager |
| 12 | 1.0000 | 32 | 0.0000  | 0.5 | j+1 > 32, then value at 32 picked |

The "averagers", which according to Wikipedia occur in all cases between the j extrema of 1 and 32, are in fact common but not absolutely "necessary".

This escalates the current issue to all types, not 2 only.

A reformulation is: follow [Hyndman, 1996] implementation in all types, even if apparently correct according to test result - it may well be the test case generated is not powerful enough to show significant differences, but other different cases could be.

**Comment by mafavaron** [comment]
[Hyndman, 1996] implemented for type=2: now, difference between R and _pbl_met_ results is zero.

The issue is not closed yet, however: all other cases must be corrected in conformity to [Hyndman, 1996].

**Comment by mafavaron** [comment]
Types 1-9 now [Hyndman, 1996]-conformant, and tested to equal expected values within single-precision machine epsilon. Issue can now be closed.
